### PR TITLE
Support existing container apps environment

### PR DESCRIPTION
**Pull request by davidkarlsen**
## Describe your changes
Support existing app service environment.

## Issue number

#37

## Checklist before requesting a review
- [X] The pr title can be used to describe what this pr did in `CHANGELOG.md` file
- [ ] I have executed pre-commit on my machine
- [x] I have passed pr-check on my machine

Thanks for your cooperation!

### CODE REVIEW

**Comment by davidkarlsen** [comment]
@lonegunmanb PTAL

File: main.tf [hunk 1]
```diff
@@ -1,3 +1,7 @@
+locals {
+  container_app_environment_id = coalesce(var.container_app_environment.id, try(azurerm_container_app_environment.container_env[0].id, null))
```
Comment (lonegunmanb) [inline main.tf]: Would the following expression be better?:

```hcl
try(var.container_app_environment.id, azurerm_container_app_environment.container_env[0].id)
```

File: outputs.tf [hunk 2]
```diff
@@ -10,5 +10,5 @@ output "container_app_fqdn" {
 
 output "container_app_ips" {
   description = "The IPs of the Latest Revision of the Container App."
-  value       = azurerm_container_app_environment.container_env.static_ip_address
+  value       = try(azurerm_container_app_environment.container_env[0].static_ip_address, null)
```
Comment (lonegunmanb) [inline outputs.tf]: Since we have a corresponding `local` now, could we just use `local.container_app_environment_id` here?

File: variables.tf [hunk 3]
```diff
@@ -4,6 +4,14 @@ variable "container_app_environment_name" {
   nullable    = false
 }
 
+variable "container_app_environment" {
+  type = object({
+    id = string
+  })
+  description = "Reference to existing container apps environment to use."
+  default     = null
```
Comment (lonegunmanb) [inline variables.tf:13]: Could we add a `validation` block here to ensure that `id` could not be `null`?

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @davidkarlsen for opening this pr! Some review comments, and could you please add a corresponding example to show how to use this new variable along with a corresponding e2e test? Thanks!

**Comment by davidkarlsen** [comment]
@lonegunmanb done, can we let it run to see how the test works?

File: variables.tf [hunk 4]
```diff
@@ -4,6 +4,18 @@ variable "container_app_environment_name" {
...
+  validation {
+    condition     = var.container_app_environment.id != null
```
Comment (lonegunmanb) [inline variables.tf]: Better be:

```hcl
condition     = var.container_app_environment == null ? true : var.container_app_environment.id != null
```

File: examples/existing-env/main.tf [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  location = var.location
+  name     = "example-container-app-${random_id.rg_name.hex}"
+}
+
+locals {
+  counting_app_name  = "counting-${random_id.container_name.hex}"
+  dashboard_app_name = "dashboard-${random_id.container_name.hex}"
+}
+
+resource "azapi_resource" "container_apps_environment" {
```
Comment (lonegunmanb) [inline examples/existing-env/main.tf]: Why we use `azapi` to create this `container_apps_environment` resource here?

File: examples/existing-env/versions.tf [hunk 6]
```diff
@@ -0,0 +1,22 @@
+terraform {
+  required_version = ">= 1.2"
+
+  required_providers {
+    azapi = {
+      source  = "Azure/azapi"
+      version = "~> 1.10.0"
```
Comment (lonegunmanb) [inline examples/existing-env/versions.tf]: This `~>` would pin `azapi` resource at `1.10.x`, is that correct?

File: examples/existing-env/versions.tf [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    }
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.11, < 4.0"
+    }
+    random = {
+      source  = "hashicorp/random"
+      version = ">= 3.0.0"
```
Comment (lonegunmanb) [inline examples/existing-env/versions.tf]: Please use `~> 3.0` to constrain the major version here, thanks!

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @davidkarlsen for the update, some review comments but almost LGTM.

File: examples/existing-env/versions.tf (hunk 6 above)
Comment (davidkarlsen) [reply examples/existing-env/versions.tf:7]: ah, it was really a copy paste due to the azurerm one missing features, but that's not really needed here - so let's not drag in that provider.

File: examples/existing-env/main.tf (hunk 5 above)
Comment (davidkarlsen) [reply examples/existing-env/main.tf]: ah, it was really a copy paste due to the azurerm one missing features, but that's not really needed here - so let's not drag in that provider.

Comment (davidkarlsen) [reply examples/existing-env/main.tf]: I'm using this while I wait for https://github.com/hashicorp/terraform-provider-azurerm/pull/23478

**Comment by davidkarlsen** [comment]
@lonegunmanb PTAL

File: outputs.tf (hunk 2 above)
Comment (lonegunmanb) [inline outputs.tf]: I just realized that now we have a problem here... A `null` is not acceptable for this module's users even this container app environment is provided by them. We need an extra data source to retrieve this brown field environment then return `static_ip_address` exported from this data source.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @davidkarlsen and I must apologize for my reckless reviews. I was failed to figure out all issues one time, sorry for that.

I have a new request, we need an extra data source to provide a valid output to module's caller.

File: outputs.tf (hunk 2 above)
Comment (davidkarlsen) [reply outputs.tf]: I can fix that, but the e2e tests seemed to fail on something else?

Comment (davidkarlsen) [reply outputs.tf]: and no worries on your reviews - they are great :)

Comment (lonegunmanb) [reply outputs.tf]: @davidkarlsen I've reran the e2e test on the main branch and it failed. I guess the provider might added a new requirement on the variables. I'll open a new pr to fix this issue, then you can rebase your branch and retry.

Comment (davidkarlsen) [reply outputs.tf]: @lonegunmanb added datasource, slightly changed how the existing env should be presented (name+rg instead of resourceId)

**Comment by davidkarlsen** [comment]
@lonegunmanb I've rebased against current main - can we test the e2e's again?

**Comment by lonegunmanb** [comment]
@davidkarlsen Please wait for #40 and #39 then you're ok to rebase again, thanks!

**Comment by lonegunmanb** [comment]
Hi @davidkarlsen #39 has been merged so you're good to rebase with the latest main branch and try again, thanks for your patience!

**Comment by davidkarlsen** [comment]
@lonegunmanb done (PS: The branch os open for collab to merge into)

**Comment by lonegunmanb** [comment]
Hi @davidkarlsen I believe that the failed e2e test should have been fixed by #42 , would you please rebase your pr with the latest `main` branch and try again? Sorry for your inconvenience.

**Comment by davidkarlsen** [comment]
@lonegunmanb done

**Review by lonegunmanb — APPROVED** [review]
Thanks @davidkarlsen , LGTM!

**Pull request merged**